qiskit/transpiler/target.py

```diff
@@ -603,32 +606,46 @@ def physical_qubits(self):
         """Returns a sorted list of physical_qubits"""
         return list(range(self.num_qubits))
 
-    def get_non_global_operation_names(self):
+    def get_non_global_operation_names(self, strict_direction=False):
         """Return the non-global operation names for the target
 
         The non-global operations are those in the target which don't apply
         on all qubits (for single qubit operations) or all multiqubit qargs
         (for multi-qubit operations).
 
+        Args:
+            strict_direction (bool): Whether to
+
         Returns:
             List[str]: A list of operation names for operations that aren't global in this target
         """
-        if self._non_global_basis is None:
-            incomplete_basis_gates = []
-            size_dict = defaultdict(int)
-            size_dict[1] = self.num_qubits
-            for qarg in self._qarg_gate_map:
-                if len(qarg) == 1:
-                    continue
-                size_dict[len(qarg)] += 1
-            for inst, qargs in self._gate_map.items():
-                qarg_sample = next(iter(qargs))
-                if qarg_sample is None:
-                    continue
-                if len(qargs) != size_dict[len(qarg_sample)]:
-                    incomplete_basis_gates.append(inst)
+        if strict_direction:
+            if self._non_global_strict_basis is not None:
+                return self._non_global_strict_basis
+            search_set = self._qarg_gate_map.keys()
+        else:
+            if self._non_global_basis is not None:
+                return self._non_global_basis
+
+            search_set = {frozenset(qarg) for qarg in self._qarg_gate_map if qarg != 1}
```

Comment:
It's just a small optimization the 1q operations are defined over `range(num_qubits)` so when we iterate over the qargs we don't need to manually count `1q` arguments and can just check later that `len(self._gate_map[qargs]) != num_qubits`) . We do this inline in the loop on L634-L636 but since we have to iterate over it here to convert to a `frozenset` anyway I figured just remove it from the search_set.

Reply:
ah, yep good catch